pkgs/top-level/all-packages.nix

```diff
@@ -15877,9 +15877,21 @@ with pkgs;
   haskell = callPackage ./haskell-packages.nix { };
 
   haskellPackages = dontRecurseIntoAttrs
-    # Prefer native-bignum to avoid linking issues with gmp
-    # GHC 9.6 rts can't be built statically with hadrian, so we need to use 9.4
-    # until 9.8 is ready
+    # Prefer `native-bignum` to avoid license issues with `gmp` (which is LGPL).
+    # Technically we could use gmp because I think Nix fulfills the requirements
+    # that it needs to be possible to relink the resulting binary with a changed `gmp`.
+    # However, `native-bignum` is the better default for static builds
+    # since it allows for the static binaries to be distributed independently of Nix
+    # (so the relinking requirement wouldn't be fulfilled necessarily).
+    #
+    # Static builds use older `ghc94` because:
+    # * Originally because GHC 9.6 rts could not be built statically with hadrian.
+    #   That was fixed in https://github.com/NixOS/nixpkgs/pull/317175
+    # * It now needs to be checked whether `ghc96` can be used.
```

Comment:
GHC 9.6 at least suffers from the same issue as 9.8. With the compilation issue gone we would  upgrade 9.6 except it would regress our ability to do TemplateHaskell which was working in previous NixOS releases.